--- a/etc/init.d/hprofile	2014-07-19 20:14:14.608007022 +0000
+++ a/etc/init.d/hprofile	2014-07-19 20:09:14.482006076 +0000
@@ -8,20 +8,16 @@
 }
 
 start() {
-	local vga=$(/usr/sbin/hpdet vga)
-	local pwr=$(/usr/sbin/hpdet power)
-	ebegin "Selecting hardware profile vga.$vga and power.$pwr"
-	/usr/sbin/hprofile vga 
-	/usr/sbin/hprofile pwr
+	profile=$(/usr/sbin/hpdet boot)
+	ebegin "Selecting hardware profile ${profile}"
+	/usr/sbin/hprofile boot
 	eend $?
 }
 
 stop() {
-	local vga=$(/usr/sbin/hpdet vga)
-	local pwr=$(/usr/sbin/hpdet power)
-	ebegin "Selecting hardware profile vga.$vga and power.$pwr"
-	/usr/sbin/hprofile vga 
-	/usr/sbin/hprofile pwr
+	profile=$(/usr/sbin/hprofile -c boot)
+	ebegin "Stopping hardware profile ${profile}"
+	/usr/sbin/hprofile -s boot
 	eend $?
 }
 
